pkg/storage/schema.go

```diff
@@ -80,20 +80,20 @@ func (sm *SchemaManager) MigrateSchemaToVersion(ctx context.Context, version uin
 }
 
 // GetSchemaVersion retrieves the current version of database schema
-func (sm *SchemaManager) GetSchemaVersion(ctx context.Context) (uint, error) {
+func (sm *SchemaManager) GetSchemaVersion(ctx context.Context) (uint, bool, error) {
 	err := sm.EnsureConnected()
 	if err != nil {
-		return 0, err
+		return 0, false, err
 	}
-	v, _, err := sm.migrator.Version()
+	v, dirty, err := sm.migrator.Version()
 	if err == migrate.ErrNilVersion {
 		// special handling for the version not being present
-		return 0, nil
+		return 0, dirty, nil
 	}
 	if err != nil {
-		return 0, err
+		return 0, dirty, err
```

Comment:
question: is it safe assumption that any error is due to db in dirty state? could it be programming bug?

Reply:
No. Dirty state is not error, it is a valid state from migration library perspective when we ask it for the schema version.
Generally `dirty`, and any other return value (like schema version) is considered undefined for error cases and should not be checked by caller if `err` is set.  This is based on Go  general convention https://blog.golang.org/error-handling-and-go - for example, square root of negative number is not 0, it is an irrational number, but Go creators are fine with returning 0 as the sqrt of negative if `err` is not `nil`, as it is caller's responsibility to check for error and ignore any return values for failed calls.  
Whether we return true of false here should not matter, i.e. we could put `true` or `false` here and if the caller's behavior changed, it would be a bug in caller code. 
We *could* use non-boolean type here and have special `undefined` state in addition of true/false, but that would probably create more problems than it solves, as the library we are wrapping does use boolean. It also would arguably be Java-in-golang. One more approach could be to abuse the pointer to boolean, and return true/false for when we "know" the result, and nil pointer when we don't, but from what I understand this would not be idiomatic go either.

Reply:
This is great, for a golang newbie thank you for sharing details